content/en/blog/_posts/2023-03-08-nodeport-dynamic-and-static-allocation.md

```diff
@@ -0,0 +1,162 @@
+---
+layout: blog
+title: "Kubernetes 1.27: Avoid Collisions Assigning ports to Nodeport Services"
+date: 2023-03-08
+slug: nodeport-dynamic-and-static-allocation
+---
+
+**Author:** xuzhenglun (Alibaba)
+
+In Kubernetes, a Service can be used to provide a unified traffic endpoint for 
+applications running on a set of Pods. Clients can use the virtual IP address provided
+by the Service for access, and Kubernetes provides load balancing for traffic accessing
+different back-end Pods, but a ClusterIP type of Service is limited to providing access to
+nodes within the cluster, while traffic from outside the cluster cannot be routed.
+One way to solve this problem is to use a NodePort type of Service, which maps a VIP's port
+to a specific port of all nodes in the cluster, thus redirecting traffic from the
+outside to the inside of the cluster.
+
+## How NodePort's ports are allocated? 
+
+When a NodePort Service is created, its corresponding port is allocated in two ways.
+
+- **Dynamic** : If the Service type is NodePort and the corresponding `service.spec.ports.nodePort`
+  is not specified, the Kubernetes control plane will automatically allocate an unused port 
+  to it at creation time.
+
+- **Static** : In addition to the dynamic auto-assignment described above, you can also
+  explicitly assign a port that is within the nodeport port range configuration.
+
+the `service.spec.ports.nodePort` must be unique for each NodePort Service across the cluster.
+Attempting to create a NodePort type of Service with a port already allocated will return an error.
+
+## Why do you need to reserve ports of NodePort Service? 
+Sometimes, you may want to have a NodePort Service running on well-known ports
+so that other components and users inside o r outside the cluster can use them.
+
+In some complex cluster deployments with a mix of K8S nodes and non-K8S nodes, it may be
+necessary to use some pre-defined ports for communication. In particular, some fundamental
+components cannot rely on the VIPs which provided by LoadBalancer services,
+because the load balancer that provides the VIPs may also rely on these fundamental components.
```

Comment:
```suggestion
because the virtual IP address mapping implementation for that cluster also relies on
these foundational components.
```

Is this right? I am not sure I have understood the message intended here.

Reply:
done